documentation/tutorial/extra/async/Async in flow.md

```diff
@@ -0,0 +1,490 @@
+            choose {
+                when(wait(delayer)) {
+                    std::cout << ++secondTicker << " second(s) has passed" << std::endl;
+                    if (secondTicker > 10) {
+                        break;
+                    }
+                    delayer = delay(1.0);
+                }
+                when(wait(exception)) {
+                    std::cout << "This should not happen." << std::endl;
+                }
+            }
+        } catch(Error& ex) {
+            std::cout << "Caught exception: " << ex.code() << " -- " << ex.what() << std::endl;
+            break;
+        }
+    }
+
+    return Void();
+}
+```
+
+See `exception.actor.cpp` for more details.
+
+### `template`d  `ACTOR`s
+
+`ACTOR`s can be `template`d for meta-programming.  To allow `ACTOR`s being transpiled properly, the keyword `ACTOR` must be placed before `template`, e.g.
+
+```C++
+ACTOR
+template <typename T>
+Future<T> coroutine();
+```
+
+An example can be found in `reportLatency.actor.cpp`.
+
+## `Promise` and `Future`
+
+`Promise`s and `Future`s play central roles of an asynchronous system. `Promise`s are used to claim a resource, that might not be immediately available, can be `wait`ed through a `Future` object. 
+
+A `Promise` can trigger multiple `Future`s that subscribes the `Promise`.  The following code serve as a  simple example of `Promise` :
+
+```C++
+const int VALUE = 42;
+Promise<int> promise;
+
+ACTOR Future<Void> waitForPromise(int id) {
+    state int value = wait(promise.getFuture());
+
+    std::cout << "waitForPromise id=" << id << " received " << value << std::endl;
+
+	return Void();
+}
+
+ACTOR Future<Void> loopForPromise() {
+    state Future<Void> delayer = delay(1.0);
+    state Future<Void> future1 = waitForPromise(1);
+    state Future<Void> future2 = waitForPromise(2);
+    state int counter = 0;
+
+    loop {
+        choose {
+            when(wait(delayer)) {
+                ++counter;
+                std::cout << counter << " second(s) has passed." << std::endl;
+                if (counter == 5) {
+                    break;
+                }
+                if (counter == 2) {
+                    promise.send(3);
+                }
+                delayer = delay(1.0);
+            }
+            when(wait(future1)) {
+                std::cout << "Future1 is done" << std::endl;
+                // It is important to mute the future1 to be Never() so it will
+                // not be triggered anymore in the loop; otherwise, since the
+                // condition will be fulfilled in the 2th second, future1 will
+                // *always* be triggered and other conditions will be ignored.
+                future1 = Never();
+            }
+            when(wait(future2)) {
+                std::cout << "Future2 is done" << std::endl;
+                future2 = Never();
+            }
+        }
+    }
+
+    return Void();
+}
+```
+
+The full example can be found in `promise.actor.cpp`
```

Comment:
Maybe also say what's the expected output?

Reply:
The expected output is intended not being included in this documentation for two reasons:

- The reader is encouraged to read the code, compile it, and see what the output would look like.
- In general, async code is not determinstic. (even this may not be true in FDB)